File: backend/app/domain/policy.py
"""The adaptation state machine.

A pure function from evidence to a decision. No database, no HTTP, no framework
— which is what lets every branch be tested in milliseconds against the captured
goldens instead of through the app.

Ported from ``frontend/src/lib/adaptation/policy.ts``. The reason strings are
reproduced exactly: they render verbatim under "Why this path", and
FRONTEND_PLAN §19 is blunt that "the reason string is what distinguishes
adapting from appearing to adapt".

No em dashes in learner-facing copy, by product decision. A colon separates a
label from its explanation; a full stop separates two complete clauses.

DECISION ORDER IS THE CONTRACT. First match wins, and the order is not
arbitrary:

1. no focus topic       — a malformed session still yields a reviewable screen
2. prerequisite_reset   — repeated misses AND a prerequisite exists
3. timed_drill          — a fast correct streak
4. worked_transfer      — the latest answer was wrong
5. guided_practice      — latest correct but slow
6. focused_practice     — the default

Note 2 before 3: a learner who has just missed twice should step back to the
prerequisite even if older answers were fast and correct. And note the
`and a prerequisite exists` guard — repeated misses on a topic with no
prerequisite falls THROUGH to worked_transfer, which is easy to get wrong.
"""

from dataclasses import dataclass
from typing import Literal

from app.domain.constants import (
    FAST_CORRECT_STREAK,
    FAST_RESPONSE_MAX_MS,
    REPEATED_MISSES,
    SLOW_RESPONSE_MIN_MS,
)
from app.domain.evidence import ObservedAnswer, SessionEvidence, TopicEvidence

InstructionalState = Literal[
    "worked_transfer",
    "guided_practice",
    "focused_practice",
    "timed_drill",
    "prerequisite_reset",
]


@dataclass(frozen=True, slots=True)
class Decision:
    state: InstructionalState
    topic: TopicEvidence
    reason: str


def _is_fast_correct(answer: ObservedAnswer) -> bool:
    return answer.correct and answer.elapsed_ms <= FAST_RESPONSE_MAX_MS


def _repeated_miss(topic: TopicEvidence) -> bool:
    recent = topic.trailing(REPEATED_MISSES)
    return len(recent) == REPEATED_MISSES and all(not a.correct for a in recent)


def _fast_correct_streak(topic: TopicEvidence) -> bool:
    recent = topic.trailing(FAST_CORRECT_STREAK)
    return len(recent) == FAST_CORRECT_STREAK and all(_is_fast_correct(a) for a in recent)


def decide(evidence: SessionEvidence) -> Decision:
    """Choose the instructional state and the topic it applies to."""
    focus = evidence.topic(evidence.focus_topic_id)

    # A malformed session should still produce a complete, reviewable screen
    # rather than an error page.
    if focus is None:
        candidates = list(evidence.topics.values())
        if not candidates:
            raise ValueError("adaptation needs at least one topic")
        fallback = candidates[0]
        return Decision(
            state="focused_practice",
            topic=fallback,
            reason=f"Starting with a {fallback.label.lower()} check.",
        )

    if _repeated_miss(focus) and focus.prerequisite_topic_id:
        prerequisite = evidence.topic(focus.prerequisite_topic_id)
        if prerequisite is not None:
            return Decision(
                state="prerequisite_reset",
                topic=prerequisite,
                reason=(
                    f"Back to {prerequisite.label.lower()} first. "
                    f"That is what {focus.label.lower()} depends on."
                ),
            )

    if _fast_correct_streak(focus):
        return Decision(
            state="timed_drill",
            topic=focus,
            reason=f"Timed drill: two fast correct {focus.label.lower()} answers.",
        )

    latest = focus.latest

    if latest is not None and not latest.correct:
        return Decision(
            state="worked_transfer",
            topic=focus,
            reason=f"More guidance: your last {focus.label.lower()} answer was wrong.",
        )

    if latest is not None and latest.correct and latest.elapsed_ms >= SLOW_RESPONSE_MIN_MS:
        return Decision(
            state="guided_practice",
            topic=focus,
            reason=(
                f"More guidance: your last {focus.label.lower()} answer "
                "was correct but slow."
            ),
        )

    reason = (
        f"Near-transfer question: your last {focus.label.lower()} answer was correct."
        if latest is not None and latest.correct
        else f"Starting with a {focus.label.lower()} check."
    )
    return Decision(state="focused_practice", topic=focus, reason=reason)

File: frontend/src/lib/adaptation/policy.ts
import 'server-only';

import type {
  AdaptationSummary,
  Block,
  InstructionalState,
  Profile,
  Quiz,
} from '../types';
import { ADAPTATION_THRESHOLDS } from './constants';
import { audioFor, explainerFor, flashcardsFor, questionFor } from './materials';
import type { ObservedAnswer, SessionEvidence, TopicEvidence } from './types';

/**
 * Behavioural adaptation fixture standing in for backend intelligence.
 *
 * This module is server-only and is called exclusively by `askFixture`, which
 * itself is reached only from a route handler's `usingFixtures` branch. It is
 * not frontend domain logic and must be replaced by a backend-owned policy when
 * that service exists (FRONTEND_PLAN.md R1/R10).
 */

export type AdaptationPlan = AdaptationSummary & { blocks: Block[] };

function last(topic: TopicEvidence): ObservedAnswer | undefined {
  return topic.recent.at(-1);
}

function trailing(topic: TopicEvidence, count: number): ObservedAnswer[] {
  return topic.recent.slice(-count);
}

function isFastCorrect(answer: ObservedAnswer): boolean {
  return answer.correct && answer.elapsed_ms <= ADAPTATION_THRESHOLDS.fastResponseMaxMs;
}

function repeatedMiss(topic: TopicEvidence): boolean {
  const recent = trailing(topic, ADAPTATION_THRESHOLDS.repeatedMisses);
  return (
    recent.length === ADAPTATION_THRESHOLDS.repeatedMisses &&
    recent.every((answer) => !answer.correct)
  );
}

function fastCorrectStreak(topic: TopicEvidence): boolean {
  const recent = trailing(topic, ADAPTATION_THRESHOLDS.fastCorrectStreak);
  return (
    recent.length === ADAPTATION_THRESHOLDS.fastCorrectStreak &&
    recent.every(isFastCorrect)
  );
}

function decisionFor(evidence: SessionEvidence): {
  state: InstructionalState;
  topic: TopicEvidence;
  reason: string;
} {
  const focus = evidence.topics[evidence.focus_topic_id];

  // A malformed fixture should still yield a complete, reviewable screen.
  if (!focus) {
    const fallback = Object.values(evidence.topics)[0];
    if (!fallback) {
      throw new Error('The adaptation fixture needs at least one topic.');
    }
    return {
      state: 'focused_practice',
      topic: fallback,
      reason: `Starting with a ${fallback.label.toLowerCase()} check.`,
    };
  }

  if (repeatedMiss(focus) && focus.prerequisite_topic_id) {
    const prerequisite = evidence.topics[focus.prerequisite_topic_id];
    if (prerequisite) {
      return {
        state: 'prerequisite_reset',
        topic: prerequisite,
        reason:
          `Back to ${prerequisite.label.toLowerCase()} first — ` +
          `that is what ${focus.label.toLowerCase()} depends on.`,
      };
    }
  }

  if (fastCorrectStreak(focus)) {
    return {
      state: 'timed_drill',
      topic: focus,
      reason: `Timed drill — two fast correct ${focus.label.toLowerCase()} answers.`,
    };
  }

  const latest = last(focus);
  if (latest && !latest.correct) {
    return {
      state: 'worked_transfer',
      topic: focus,
      reason: `More guidance — your last ${focus.label.toLowerCase()} answer was wrong.`,
    };
  }

  if (
    latest?.correct &&
    latest.elapsed_ms >= ADAPTATION_THRESHOLDS.slowResponseMinMs
  ) {
    return {
      state: 'guided_practice',
      topic: focus,
      reason:
        `More guidance — your last ${focus.label.toLowerCase()} answer ` +
        'was correct but slow.',
    };
  }

  return {
    state: 'focused_practice',
    topic: focus,
    reason: latest?.correct
      ? `Near-transfer question — your last ${focus.label.toLowerCase()} answer was correct.`
      : `Starting with a ${focus.label.toLowerCase()} check.`,
  };
}

function quizFor(
  topicId: string,
  state: InstructionalState,
  profile: Profile,
): Quiz {
  const kind =
    state === 'timed_drill'
      ? 'timed'
      : state === 'focused_practice'
        ? 'transfer'
        : 'guided';

  return {
    type: 'quiz',
    // The approved profile invariant reserves visible timers for `strong`.
    // Other profiles still receive the compressed question selected by the
    // timed-drill state, but without a countdown.
    timer_seconds:
      profile === 'strong'
        ? state === 'timed_drill'
          ? ADAPTATION_THRESHOLDS.compressedTimerSeconds
          : ADAPTATION_THRESHOLDS.standardTimerSeconds
        : null,
    questions: [questionFor(topicId, kind)],
  };
}

function instructionalCore(
  topicId: string,
  state: InstructionalState,
  profile: Profile,
): Block[] {
  const quiz = quizFor(topicId, state, profile);

  if (
    state === 'worked_transfer' ||
    state === 'guided_practice' ||
    state === 'prerequisite_reset'
  ) {
    // Faded only for `guided_practice` — the learner answered correctly but
    // slowly, so they can already do this and completing the example beats
    // reading it. A learner who got it wrong, or is back on a prerequisite,
    // still gets the whole thing.
    return [explainerFor(topicId, state === 'guided_practice'), quiz];
  }

  return [quiz];
}

/**
 * Choose and compose the next screen from observed behaviour plus a delivery
 * preference. No profile is a pre-authored screen: it changes ordering,
 * modality and density around the instructional state selected above.
 */
export function policy(evidence: SessionEvidence, profile: Profile): AdaptationPlan {
  const decision = decisionFor(evidence);
  let core = instructionalCore(decision.topic.topic_id, decision.state, profile);

  // `rusty` must still open with a segmented worked example even after strong
  // evidence. The invariant will clamp this raw set to that first block.
  if (profile === 'rusty' && core[0]?.type !== 'explainer_card') {
    core = [explainerFor(decision.topic.topic_id), ...core];
  }

  let blocks: Block[];
  switch (profile) {
    case 'rusty':
      blocks = [...core, evidence.roadmap, evidence.progress];
      break;
    case 'time_poor':
      blocks = [evidence.roadmap, ...core, evidence.progress];
      break;
    case 'hands_free':
      blocks = [audioFor(decision.topic.topic_id), ...core, flashcardsFor(decision.topic.topic_id)];
      break;
    case 'strong':
      blocks = [...core, evidence.roadmap, evidence.progress];
      break;
  }

  return {
    state: decision.state,
    topic_id: decision.topic.topic_id,
    reason: decision.reason,
    blocks: blocks.slice(0, 5),
  };
}

